File: 백준/Gold/11066. 파일 합치기/파일 합치기.java
import java.io.*;
import java.util.StringTokenizer;

class Main {
    static int[][] costs;
    static int[] cumSum;
    static int k;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int totalCase = Integer.parseInt(br.readLine());

        while (totalCase-- > 0) {
            k = Integer.parseInt(br.readLine());
            costs = new int[k + 1][k + 1];
            cumSum = new int[k + 1];

            StringTokenizer st = new StringTokenizer(br.readLine());

            for (int i = 1; i <= k; i++) {
                cumSum[i] = cumSum[i - 1] + Integer.parseInt(st.nextToken());
            }

            bw.append(String.valueOf(solve()));
            bw.newLine();
        }

        bw.flush();
        bw.close();
        br.close();
    }

    static int solve() {
        for (int size = 1; size < k; size++) {
            for (int s = 1; s <= k - size; s++) {
                int e = s + size;
                int cost = Integer.MAX_VALUE;

                for (int mid = s; mid < e; mid++) {
                    cost = Math.min(cost, costs[s][mid] + costs[mid + 1][e]);
                }

                costs[s][e] = cost + cumSum[e] - cumSum[s - 1];
            }
        }

        return costs[1][k];
    }
}

File: 백준/Gold/11066. 파일 합치기/파일 합치기.kt
import java.util.*

lateinit var memo: Array<IntArray>
lateinit var sums: IntArray

fun main() {
    val br = System.`in`.bufferedReader()
    val bw = System.out.bufferedWriter()
    val totalCast = br.readLine().toInt()

    repeat(totalCast) {
        val k = br.readLine().toInt()

        memo = Array(k + 1) { IntArray(k + 1) }
        sums = IntArray(k + 1)

        val st = StringTokenizer(br.readLine())

        for (i in 1..k) {
            sums[i] = sums[i - 1] + st.nextToken().toInt()
        }

        bw.append("${trace(1, k)}\n")
    }

    bw.flush()
    bw.close()
    br.close()
}

fun trace(s: Int, e: Int): Int {
    if (s == e) return 0

    if (memo[s][e] != 0) return memo[s][e]

    var cost = Int.MAX_VALUE

    for (mid in s until e) {
        cost = minOf(cost, trace(s, mid) + trace(mid + 1, e))
    }

    memo[s][e] = cost + sums[e] - sums[s - 1]

    return memo[s][e]
}